sel/src/Sel/Hashing/Short.hs

```diff
@@ -0,0 +1,297 @@
+{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE DerivingVia #-}
+{-# LANGUAGE TypeApplications #-}
+
+-- |
+--
+-- Module: Sel.Hashing.Short
+-- Description: Short input hashing with the SipHash-2-4 algorithm
+-- Copyright: (C) Hécate Moonlight 2023
+-- License: BSD-3-Clause
+-- Maintainer: The Haskell Cryptography Group
+-- Portability: GHC only
+module Sel.Hashing.Short
+  ( -- * Introduction
+    -- $introduction
+    ShortHash
+
+    -- ** Short-input Hashing
+  , hashByteString
+  , hashText
+
+    -- *** Conversion
+  , shortHashToBinary
+  , shortHashToHexText
+  , shortHashToHexByteString
+
+    -- ** Short Hash Key
+  , ShortHashKey
+  , newKey
+
+    -- *** Conversion
+  , shortHashKeyToBinary
+  , shortHashKeyToHexText
+  , shortHashKeyToHexByteString
+  , binaryToShortHashKey
+  , hexTextToShortHashKey
+  , hexByteStringToShortHashKey
+
+    -- ** Errors
+  , ShortHashingException (..)
+  )
+where
+
+import Control.Exception (throw)
+import Control.Monad (void, when)
+import Data.ByteString (StrictByteString)
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Base16 as Base16
+import qualified Data.ByteString.Internal as BS
+import qualified Data.ByteString.Unsafe as BS
+import Data.Text (Text)
+import qualified Data.Text.Encoding as Text
+import qualified Data.Text.Lazy.Builder as Builder
+import Foreign hiding (void)
+import Foreign.C (CSize, CUChar, CULLong)
+import GHC.Exception (Exception)
+import GHC.IO.Handle.Text (memcpy)
+import System.IO.Unsafe (unsafeDupablePerformIO)
+
+import Data.Text.Display
+import LibSodium.Bindings.ShortHashing
+  ( cryptoShortHashSipHashX24Bytes
+  , cryptoShortHashSipHashX24KeyBytes
+  , cryptoShortHashX24
+  , cryptoShortHashX24KeyGen
+  )
+import Sel.Internal
+
+-- $introduction
+--
+-- This module provides an API for performant short-input hashing,
+-- backed by the [SipHash-2-4](https://en.wikipedia.org/wiki/SipHash) algorithm.
+--
+-- Short-input hashing functions have a variety of use-cases, such as:
+--
+-- * Hash Tables
+-- * Probabilistic data structures, such as Bloom filters
+-- * Integrity checking in interactive protocols
+
+-- | A 128-bit hash of a short input, of size 'cryptoShortHashSipHashX24Bytes'
+--
+-- @since 0.0.1.0
+newtype ShortHash = ShortHash (ForeignPtr CUChar)
+  deriving stock
+    ( Show
+      -- ^ @since 0.0.1.0
+    )
```

Comment:
This'll just spew an address. Is that what we want for the `Show` instance?

Reply:
Yes I know, I thought for a bit that it would be relevant to have `Show` instance display the internal of the structure, and Display show a more user-facing representation (hexadecimal), but perhaps it's not really a problem I should concern myself with and basically have Show == Display?

Reply:
Alrighty